internal/api/promote_subscribers_v1alpha1.go

```diff
@@ -0,0 +1,90 @@
+package api
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"connectrpc.com/connect"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kubev1alpha1 "github.com/akuity/kargo/api/v1alpha1"
+	typesv1alpha1 "github.com/akuity/kargo/internal/api/types/v1alpha1"
+	"github.com/akuity/kargo/internal/kargo"
+	"github.com/akuity/kargo/internal/logging"
+	svcv1alpha1 "github.com/akuity/kargo/pkg/api/service/v1alpha1"
+	v1alpha1 "github.com/akuity/kargo/pkg/api/v1alpha1"
+)
+
+func (s *server) PromoteSubscribers(
+	ctx context.Context,
+	req *connect.Request[svcv1alpha1.PromoteSubscribersRequest],
+) (*connect.Response[svcv1alpha1.PromoteSubscribersResponse], error) {
+	if err := validateProjectAndStageNonEmpty(req.Msg.GetProject(), req.Msg.GetStage()); err != nil {
+		return nil, err
+	}
+	if err := s.validateProject(ctx, req.Msg.GetProject()); err != nil {
+		return nil, err
+	}
+	stage, err := getStage(ctx, s.client, req.Msg.GetProject(), req.Msg.GetStage())
+	if err != nil {
+		return nil, err
+	}
+	if err = validateFreightExists(req.Msg.GetFreight(), stage.Status.History); err != nil {
+		return nil, err
+	}
+
+	subscribers, err := s.findStageSubscribers(ctx, stage)
+	if err != nil {
+		return nil, err
+	}
+	if len(subscribers) == 0 {
+		return nil, connect.NewError(connect.CodeNotFound, fmt.Errorf("Stage %q has no subscribers", req.Msg.GetStage()))
+	}
```

Comment:
I have no strong opinion here, but I at last wondered if this should just be an implicit no-op instead of an error. I'm ok with either, but thought it was worth bringing up.

Reply:
It's a good call out. I actually didn't have this check before, but I added it last night minute because I felt it was more than likely a user error if they were promoting something to subscribers and there was none. We might need to introduce a flag to allow this, if the use case does pops up (I can't think of one).

Reply:
Not that it stops us from figuring out the right behavior here, but maybe the UI doesn't need to show the promote to subscribers button if there are no subscribers?

Reply:
I agree UI could do a lot to make this mostly a non-issue. I'm happy to remove this if you want. I also don't have a strong opinion